File: mariadb-api/src/Controllers/OrganisationAnnouncementsController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\AnnouncementService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/**
 * Ported from php-api's Controllers/OrganisationAnnouncementsController.php. Org-Admin-only
 * management of the calling org's own Announcements/Disruption Notices — mirrors PortfolioController's
 * shape (a dedicated controller nested under api/organisations/me, rather than a method on
 * OrganisationsController) since this is its own distinct CRUD surface. Every write is always
 * Scope="org" + OrganisationId/CreatedByUserId re-derived from the caller's own JWT claims, never
 * client-supplied — see AnnouncementService's own doc comment.
 */
final class OrganisationAnnouncementsController extends BaseController
{
    private function service(): AnnouncementService
    {
        return new AnnouncementService(Database::connection());
    }

    public function list(Request $request, Response $response): Response
    {
        return $this->json($response, $this->service()->listForOrg($this->callerOrgId($request)));
    }

    public function create(Request $request, Response $response): Response
    {
        $body = $this->body($request);
        return $this->json($response, $this->service()->create($this->callerOrgId($request), $this->callerUserId($request), $body));
    }

    public function update(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $result = $this->service()->update($this->callerOrgId($request), $args['announcementId'], $body);
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function delete(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->service()->delete($this->callerOrgId($request), $args['announcementId']);
        return $deleted ? $this->noContent($response) : $this->notFound($response);
    }
}

File: api/Enkl.Api/Controllers/OrganisationAnnouncementsController.cs
using Enkl.Api.Auth;
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

/// <summary>
/// Org-Admin-only management of the calling org's own Announcements/Disruption Notices — mirrors
/// PortfolioController's shape (a dedicated controller nested under api/organisations/me, rather than
/// a method on OrganisationsController) since this is its own distinct CRUD surface. Every write is
/// always Scope="org" + OrganisationId/CreatedByUserId re-derived from User.OrgId()/User.UserId(),
/// never client-supplied — see AnnouncementService's own doc comment.
/// </summary>
[ApiController]
[Authorize(Policy = "OrgAdmin")]
[Route("api/organisations/me/announcements")]
public class OrganisationAnnouncementsController : ControllerBase
{
    private readonly AnnouncementService _announcements;

    public OrganisationAnnouncementsController(AnnouncementService announcements)
    {
        _announcements = announcements;
    }

    [HttpGet]
    public async Task<IActionResult> List()
    {
        return Ok(await _announcements.ListForOrgAsync(User.OrgId()));
    }

    [HttpPost]
    public async Task<IActionResult> Create(CreateAnnouncementRequest request)
    {
        return Ok(await _announcements.CreateAsync(User.OrgId(), User.UserId(), request));
    }

    [HttpPut("{announcementId:guid}")]
    public async Task<IActionResult> Update(Guid announcementId, UpdateAnnouncementRequest request)
    {
        var result = await _announcements.UpdateAsync(User.OrgId(), announcementId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{announcementId:guid}")]
    public async Task<IActionResult> Delete(Guid announcementId)
    {
        var deleted = await _announcements.DeleteAsync(User.OrgId(), announcementId);
        return deleted ? NoContent() : NotFound();
    }
}
